File: 0322-coin-change/0322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if (coins == null || coins.length == 0 || amount == 0) return 0;

        int n = coins.length;

        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1);
        dp[0] = 0;

        for (int coin: coins) {
            for (int j = coin; j <= amount; j++) {
                dp[j] = Math.min(dp[j], dp[j - coin] + 1);
            }
        }

        return dp[amount] > amount ? -1 : dp[amount];
    }
}

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if not coins or amount == 0:
            return 0

        memo = {}

        def helper(amount: int, index: int) -> int:
            if amount == 0:
                return 0
            
            if amount < 0 or index == len(coins):
                return float('inf')

            if (amount, index) in memo:
                return memo[(amount, index)]

            case0 = helper(amount, index + 1) 

            case1 = helper(amount - coins[index], index)
            if case1 != float('inf'):
                case1 += 1

            memo[(amount, index)] = min(case0, case1)

            return memo[(amount, index)]

        result = helper(amount, 0)

        return -1 if result == float('inf') else result


            